### PR TITLE
Algorithmic Directional Lighting

### PR DESCRIPTION
**Pull request by Lohikar**
This is similar to #1722, but it uses a bit of math to allow explicit setting of the angle of light.
Pretty much complete, although no lights have been updated to use directional lighting.

Issues:
- Facing a wall with a directional light results in darkness - light is not emitted on own tile.
- ~~Changing the direction you're facing does not update the light - you must move.~~ Fixed.

### CODE REVIEW

File: code/modules/lighting/lighting_source.dm [hunk 1]
```diff
@@ -222,6 +237,71 @@
...
+		if (EAST) // the light will face this way by default.
+			translated_angle = light_angle
+
+		if (WEST)
+			translated_angle = light_angle + HALF_CIRCLE
+
+	limit_a_x = POLAR_TO_CART_X(light_range + 10, translated_angle)	// Convert our angle + range into a vector.
+	limit_a_y = POLAR_TO_CART_Y(light_range + 10, translated_angle)	// 10 is an arbitrary number, yes.
+	limit_b_x = limit_a_x
+	limit_b_y = POLAR_TO_CART_Y(light_range + 10, 360 - translated_angle)	// limit_b is limit_a reflected over the x-axis.
```
Comment (Lohikar) [inline code/modules/lighting/lighting_source.dm]: I'm pretty sure the trig around here is what's causing the lighting anomalies.

File: code/modules/lighting/lighting_source.dm [hunk 2]
```diff
@@ -6,10 +6,13 @@
...
 	var/turf/source_turf     // The turf under the above.
-	var/light_power    // Intensity of the emitter light.
-	var/light_range      // The range of the emitted light.
-	var/light_color    // The colour of the light, string, decomposed by parse_light_color()
+	var/light_power    	// Intensity of the emitter light.
+	var/light_range     // The range of the emitted light.
+	var/light_color    	// The colour of the light, string, decomposed by parse_light_color()
 	var/light_uv		// The intensity of UV light, between 0 and 255.
+	var/light_angle		// The light's emission angle, in degrees.
+	var/light_dir = EAST		// The direction the light is facing.
```
Comment (skull132) [inline code/modules/lighting/lighting_source.dm]: This is unnecessary. Why not just fetch top_atom.dir every time?

Comment (Lohikar) [inline code/modules/lighting/lighting_source.dm]: It doesn't get any information from the top atom yet, that was just for debugging.

**Comment by Lohikar** [comment]
Updated screenshots:

File: code/modules/lighting/lighting_verbs.dm [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+/client/proc/lighting_clear_overlay(turf/T in world)
+	set category = "Lighting"
+	set name = "Clear Overlay"
+	set desc = "Clears a lighting overlay for a turf if it has one."
+
+	CHECK_RIGHTS(R_DEBUG|R_DEV)
+		
+	if (!T.lighting_overlay)
+		src << "That turf doesn't have a lighting overlay."
```
Comment (skull132) [inline code/modules/lighting/lighting_verbs.dm]: Needs a server log. Devs removing lighting overlays to see stuff they shouldn't be able to when.

File: code/modules/lighting/lighting_verbs.dm [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	T.reconsider_lights()
+
+/client/proc/lighting_build_overlay(turf/T in world)
+	set category = "Lighting"
+	set name = "Build Overlay"
+	set desc = "Builds a lighting overlay for a turf if it does not have one."
+
+	CHECK_RIGHTS(R_DEBUG|R_DEV)
+
+	if (T.lighting_overlay)
```
Comment (skull132) [inline code/modules/lighting/lighting_verbs.dm]: Log.

File: code/modules/lighting/lighting_verbs.dm [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+		return
+
+	log_and_message_admins("has flushed the lighting processor queues.")
+	lighting_update_lights = list()
+	lighting_update_corners = list()
+	lighting_update_overlays = list()
+
+/client/proc/lighting_reconsider_target(turf/T in world)
+	set category = "Lighting"
+	set name = "Reconsider Visibility"
```
Comment (skull132) [inline code/modules/lighting/lighting_verbs.dm]: Log

File: code/modules/lighting/lighting_verbs.dm [hunk 6]
```diff
@@ -0,0 +1,102 @@
+#define CHECK_RIGHTS(PERM)     \
```
Comment (skull132) [inline code/modules/lighting/lighting_verbs.dm]: Unnecessary. check_rights already displays an error message. If it doesn't by default, then there's a boolean in its params you can touch to make it do so.

**Pull request merged**